File: LeetCode/DP/Golang/Leetcode940-不同的子序列.go
package sbgo

func distinctSubseqII(s string) int {
	MOD, n := int(1e9)+7, len(s)
	f := make([]int, 26)
	f[s[0]-'a'] = 1
	sum := func(arr []int) int {
		res := 0
		for _, val := range arr {
			res = (res + val) % MOD
		}
		return res
	}
	for i := 1; i < n; i++ {
		rec := 0
		rec = sum(f) + 1
		f[s[i]-'a'] = rec
	}
	return sum(f)
}

File: LeetCode/DP/Python/Leetcode940-不同的子序列.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        f,P = [0] * 26, int(1e9 + 7)
        f[ord(s[0]) - ord('a')] = 1
        for i in range(1, len(s)):
            rec = 0
            rec = (sum(f) + 1) % P
            f[ord(s[i]) - ord('a')] = rec
        return sum(f) % P